From bf3f571eb96753881d2bcb971ae3f811425918b0 Mon Sep 17 00:00:00 2001
From: dianlujitao <dianlujitao@lineageos.org>
Date: Fri, 2 Feb 2018 18:59:59 +0800
Subject: [PATCH 090/178] input: fingerprint: fpc1020: Respect capacitive keys
 state

 * Implement a key handler with the simple purpose of filtering
   the key events sent by the fingerprint HAL.
 * Only report key event when capacitive keys enabled.
 * Input handler taken from bgcngm's implementation for Xiaomi msm8996.

Change-Id: I091eb514fee291b7625193496c217c0b966fa518
---
 .../fingerprint/fpc/fpc1020_platform_tee.c    | 87 ++++++++++++++++++-
 1 file changed, 86 insertions(+), 1 deletion(-)

diff --git a/drivers/input/fingerprint/fpc/fpc1020_platform_tee.c b/drivers/input/fingerprint/fpc/fpc1020_platform_tee.c
index 719b3093a288..db83d97732b6 100644
--- a/drivers/input/fingerprint/fpc/fpc1020_platform_tee.c
+++ b/drivers/input/fingerprint/fpc/fpc1020_platform_tee.c
@@ -40,6 +40,13 @@
 #include <linux/fb.h>
 #include <linux/mdss_io_util.h>
 
+#ifdef CONFIG_TOUCHSCREEN_COMMON
+#include <linux/input.h>
+#include <linux/input/tp_common.h>
+#endif
+
+#define FPC1020_NAME "fpc1020"
+
 #define FPC_TTW_HOLD_TIME 2000
 #define FP_UNLOCK_REJECTION_TIMEOUT (FPC_TTW_HOLD_TIME - 500)
 
@@ -100,6 +107,9 @@ struct fpc1020_data {
 	bool wait_finger_down;
 	struct work_struct work;
 	bool proximity_state; /* 0:far 1:near */
+#ifdef CONFIG_TOUCHSCREEN_COMMON
+	struct input_handler input_handler;
+#endif
 };
 
 static irqreturn_t fpc1020_irq_handler(int irq, void *handle);
@@ -733,6 +743,68 @@ static const struct file_operations proc_file_fpc_ops = {
 };
 #endif
 
+#ifdef CONFIG_TOUCHSCREEN_COMMON
+static int input_connect(struct input_handler *handler,
+		struct input_dev *dev, const struct input_device_id *id) {
+	int rc;
+	struct input_handle *handle;
+	struct fpc1020_data *fpc1020 =
+		container_of(handler, struct fpc1020_data, input_handler);
+
+	if (!strstr(dev->name, "uinput-fpc"))
+		return -ENODEV;
+
+	handle = kzalloc(sizeof(struct input_handle), GFP_KERNEL);
+	if (!handle)
+		return -ENOMEM;
+
+	handle->dev = dev;
+	handle->handler = handler;
+	handle->name = FPC1020_NAME;
+	handle->private = fpc1020;
+
+	rc = input_register_handle(handle);
+	if (rc)
+		goto err_input_register_handle;
+
+	rc = input_open_device(handle);
+	if (rc)
+		goto err_input_open_device;
+
+	return 0;
+
+err_input_open_device:
+	input_unregister_handle(handle);
+err_input_register_handle:
+	kfree(handle);
+	return rc;
+}
+
+static bool input_filter(struct input_handle *handle, unsigned int type,
+		unsigned int code, int value)
+{
+	if (code == KEY_HOME) {
+		return !capacitive_keys_enabled;
+	}
+	return false;
+}
+
+static void input_disconnect(struct input_handle *handle)
+{
+	input_close_device(handle);
+	input_unregister_handle(handle);
+	kfree(handle);
+}
+
+static const struct input_device_id ids[] = {
+	{
+		.flags = INPUT_DEVICE_ID_MATCH_EVBIT,
+		.evbit = { BIT_MASK(EV_KEY) },
+	},
+	{ },
+};
+#endif
+
 static int fpc1020_probe(struct platform_device *pdev)
 {
 	struct device *dev = &pdev->dev;
@@ -807,6 +879,19 @@ static int fpc1020_probe(struct platform_device *pdev)
 
 	wake_lock_init(&fpc1020->ttw_wl, WAKE_LOCK_SUSPEND, "fpc_ttw_wl");
 
+#ifdef CONFIG_TOUCHSCREEN_COMMON
+	fpc1020->input_handler.filter = input_filter;
+	fpc1020->input_handler.connect = input_connect;
+	fpc1020->input_handler.disconnect = input_disconnect;
+	fpc1020->input_handler.name = FPC1020_NAME;
+	fpc1020->input_handler.id_table = ids;
+	rc = input_register_handler(&fpc1020->input_handler);
+	if (rc) {
+		dev_err(dev, "failed to register key handler\n");
+		goto exit;
+	}
+#endif
+
 	rc = sysfs_create_group(&dev->kobj, &attribute_group);
 	if (rc) {
 		dev_err(dev, "could not create sysfs\n");
@@ -855,7 +940,7 @@ MODULE_DEVICE_TABLE(of, fpc1020_of_match);
 
 static struct platform_driver fpc1020_driver = {
 	.driver = {
-		.name	= "fpc1020",
+		.name	= FPC1020_NAME,
 		.owner	= THIS_MODULE,
 		.of_match_table = fpc1020_of_match,
 	},
-- 
2.49.0

